ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -57,141 +51,130 @@ public CloudStorageCompactor(CloudDestination cloudDestination, CloudConfig clou
+    if (mainThread != null) {
+      logger.info("[COMPACT] Interrupting thread {}", mainThread.get().getName());
+      mainThread.get().interrupt();
     }
   }
 
   /**
    * @return whether the compactor is shutting down.
    */
   boolean isShutDown() {
-    return shutDown.get();
+    return this.executorService.isShutdown();
   }
 
   /**
    * Purge the inactive blobs in all managed partitions.
    * @return the total number of blobs purged.
    */
   public int compactPartitions() {
-    if (partitions.isEmpty()) {
-      logger.info("Skipping compaction as no partitions are assigned.");
-      return 0;
-    }
-    logger.info("Starting compaction on {} assigned partitions", partitions.size());
-
-    List<PartitionId> partitionSnapshot = new ArrayList<>(partitions);
     long compactionStartTime = System.currentTimeMillis();
-    long timeToQuit = System.currentTimeMillis() + compactionTimeLimitMs;
-    int compactionInProgress = 0;
-    doneLatch.set(new CountDownLatch(1));
     int totalBlobsPurged = 0;
-    int compactedPartitionCount = 0;
-    try {
-      while (!partitionSnapshot.isEmpty()) {
-        while (compactionInProgress < numThreads) {
-          if (partitionSnapshot.isEmpty()) {
-            break;
-          }
-          PartitionId partitionId = partitionSnapshot.remove(0);
-          executorCompletionService.submit(() -> compactPartition(partitionId));
-          compactionInProgress++;
-        }
-        try {
-          totalBlobsPurged += executorCompletionService.take().get();
-          compactedPartitionCount++;
-        } catch (ExecutionException ex) {
-          vcrMetrics.compactionFailureCount.inc();
-        }
-        compactionInProgress--;
-        if (System.currentTimeMillis() >= timeToQuit) {
-          logger.info("Compaction terminated due to time limit exceeded.");
-          break;
-        }
-        if (isShutDown()) {
-          logger.info("Compaction terminated due to shut down.");
-          break;
-        }
-      }
-      while (compactionInProgress > 0) {
-        try {
-          totalBlobsPurged += executorCompletionService.take().get();
-          compactedPartitionCount++;
-        } catch (ExecutionException ex) {
-          vcrMetrics.compactionFailureCount.inc();
-        }
-        compactionInProgress--;
+    HashMap<String, Future<Integer>> compactionTasks = new HashMap<>();
+      /*
+        Just submit the jobs and let the executor service handle the assignment and scheduling.
+        Note that partitions can be updated between compaction cycles when a new replica is added or an old one removed.
+        However, I do not know the behavior when a Collection is modified while it is being iterated.
+       */
+    for (PartitionId partitionId: partitions) {
+      try {
+        String partitionIdStr = partitionId.toPathString();
+        Future<Integer> future = executorService.submit(() -> cloudDestination.compactPartition(partitionIdStr));
```

Comment:
If compactPartitions itself is running in a background thread, just wondering on why do we need to run `cloudDestination.compactPartition` via separate thread again?

Reply:
For parallelism and faster compaction.

Reply:
Also, prev code did the same.